File: src/uva/Leetcode00020.java
import java.util.*;

public class Leetcode00020 {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Map<Character,Character> m = new HashMap<>();
        m.put('(',')');
        m.put('[',']');
        m.put('{','}');        
        for (int i = 0;i < s.length();i++) {
            char next = s.charAt(i);
            System.out.println(next);
            if (m.keySet().contains(next)) {
                stack.push(next);
            } else {
                if (stack.isEmpty()||next != m.get(stack.pop())) {
                    return false;
                }
            }
        }
        return stack.size() == 0;
    }
}

File: src/main/Leetcode00020.swift
class Stack<T> {
    var items = [T]()
    func push(i : T) {
        items.append(i)
    }
    func pop() -> T {
        return items.removeLast()
    }
    func count() -> Int {
        return items.count
    }
}

class Solution {
    func isValid(s: String) -> Bool {
        let m : [Character:Character] = [
            "(":")",
            "{":"}",
            "[":"]",
        ]
        var stack = Stack<Character>()
        for c in s.characters {
            switch (c) {
                case let value where m.keys.contains(c):
                    stack.push(c)
                case let value where m.values.contains(c):
                    if stack.count() == 0 || m[stack.pop()]! != c {
                        return false
                    }
                default:
                    return false
            }
        }
        return stack.count() == 0
    }
}
